src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,162 @@
 import java.util.List;
 
 public class MyLinkedList<T> implements MyLinkedListInterface<T> {
-    @Override
-    public void add(T value) {
+    private static final int ZERO_SIZE = 0;
+    private static final int HALF_SIZE = 2;
+    private Node<T> head;
+    private Node<T> tail;
+    private int size;
+
+    private static class Node<T> {
+        private T data;
+        private Node<T> next;
+        private Node<T> prev;
+
+        Node(T data) {
+            this.data = data;
+        }
     }
 
     @Override
-    public void add(T value, int index) {
+    public void add(T element) {
+        Node<T> newNode = new Node<>(element);
+        if (tail == null) {
+            head = tail = newNode;
+        } else {
+            tail.next = newNode;
+            newNode.prev = tail;
+            tail = newNode;
+        }
+        size++;
     }
 
     @Override
-    public void addAll(List<T> list) {
+    public void add(T element, int index) {
+        if (index < ZERO_SIZE || index > size) {
+            throw new IndexOutOfBoundsException("Index: " + index
+                    + ", Size: " + size);
+        }
+        Node<T> newNode = new Node<>(element);
+        if (index == size) {
+            add(element);
+        } else if (index == ZERO_SIZE) {
+            newNode.next = head;
+            if (head != null) {
+                head.prev = newNode;
+            }
+            head = newNode;
+            if (size == ZERO_SIZE) {
+                tail = head;
+            }
+            size++;
+        } else {
+            Node<T> current = getNode(index);
+            Node<T> previous = current.prev;
+            previous.next = newNode;
+            newNode.prev = previous;
+            newNode.next = current;
+            current.prev = newNode;
+            size++;
+        }
     }
 
     @Override
     public T get(int index) {
-        return null;
+        return getNode(index).data;
     }
 
     @Override
-    public T set(T value, int index) {
-        return null;
+    public T set(T element, int index) {
+        Node<T> node = getNode(index);
+        T oldValue = node.data;
+        node.data = element;
+        return oldValue;
     }
 
     @Override
     public T remove(int index) {
```

Comment:
Create a separate method `unlink(Node node)`. This logic will be used in two methods: when we remove from the list by index and by value. Let's create a method that will take a Node that needs to be removed and unlink it. As a result, we will call it from both methods when Node is found.